walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,86 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> overlapEvents = new ArrayList<Event>();
+    List<TimeRange> overlapRanges = new ArrayList<TimeRange>();
+    List<TimeRange> noOverlapRanges = new ArrayList<TimeRange>();
+    List<TimeRange> possibleRanges = new ArrayList<TimeRange>();
+    int requestDuration = (int) request.getDuration();
+
+    if (requestDuration > TimeRange.WHOLE_DAY.duration()) {
+      return possibleRanges;
+    }
+   
+    for (String attendee : request.getAttendees()) {
+      for (Event event : events) {
+        if (event.getAttendees().contains(attendee)) {
+          overlapEvents.add(event);
+        }
+      }
+    }
+
+    if (overlapEvents.isEmpty()) {
+      possibleRanges.add(TimeRange.WHOLE_DAY);
+      return possibleRanges;
+    }
+
+    for (Event event : overlapEvents) {
+      overlapRanges.add(event.getWhen());
+    }
+
+    Collections.sort(overlapRanges, TimeRange.ORDER_BY_START);
+  
+    int i = 0;
+    while (i < overlapRanges.size()) {
```

Comment:
The goal is to merge sorted time ranges, right? I want to propose the following tweak on your algorithm here which I feel is simpler:

```java
for (TimeRange range : overlapRanges) {
  if (noOverlapRanges.isEmpty()) {
    noOverlapRanges.add(range);
  } else {
    TimeRange prevRange = noOverlapRanges.get(noOverlapRanges.size() - 1);
    // check if range overlaps prevRange, and modify the last element or append as necessary
  }
}
```